File: add-and-search-word/add-and-search-word-data.py
#!/usr/bin/env python
class TrieNode(object):
    def __init__(self):
        self.is_word = False
        self. leaves = {}
        
class WordDictoinary(object):
    def __init__(self):
        self.root = TrieNode()
        
    def addWord(self, word):
        curr = self.root
        for c in word:
            if not c in curr.leaves:
                curr.leaves[c] = TrieNode()
                
        curr.is_word = True
    
    def search(self, word):
       return self.searchHelper(word, 0, self.root)
    
    def searchHelper(self, word, start, curr):
        if start == len(word):
            return curr.is_word
        
        if word[start] in curr.leaves:
            return self.searchHelper(word, start + 1, curr.leaves[word[start]])
            
        elif word[start] == '.':
            for c in curr.leaves:
                if self.searchHelper(word, start + 1, curr.leaves[c]):
                    return True
                    
        return False
        

File: add-and-search-word/add-and-search-word_data.cpp
#include<string.h>
#include<iostream>
#include<string>
using namespace std;

const int MAX_CHARS = 26;

/*Trie Node */
class TrieNode {
    public: 
        TrieNode():isWord(false) {
            memset(children, 0, sizeof(children));
        }
        TrieNode* & operator[](char idx) {
            int i = (idx - 'a') % MAX_CHARS;
            return children[i];
        }
        TridNode* & operator[](int idx) {
            int i = idx % MAX_CHARS;
            return children[i];
        }
        bool isWord;
    private:
        TrieNode* children[MAX_CHARS];
};

/*Trie Tree*/
class TrieTree {
    public: 
        TrieTree():root(new TrieNode()) {}
        ~TrieTree(){freeTree(root);}
        
        void put(string &s) {
            TrieNode* node = root;
            for (int i = 0; i < s.size(); i++) {
                if ((*node)[s[i]] == NULL) {
                    (*node)[s[i]] = new TrieNode();
                }
                node = (*node)[s[i]];
            }
            node->isWord = true;
        }
        bool search(string &s) {
            return get(s, this->root);
        }
    private:
        bool get(string &s, TrieNode* root, int idx=0) {
            TrieNode* node = root;
            for(int i = idx; i < s.size(); i++){
                if(s[i] != '.') {
                    node = (*node)[s[i]]
                    if(node == NULL) return false;
                } else {
                    for (int j = 0; j < MAX_CHARS; j++) {
                        TrieNode *p = (*node)[j];
                        if (p == NULL) {
                            continue;
                        }
                        
                        if(i < s.size() - 1) {
                            if(this->get(s, p, i + 1)) {
                                return true;
                            }
                            continue;
                        }
                        if(p->isWord) {
                            return true;
                        }
                    }
                    return false;
                }
            }
            return node->isWord;
            
        }
        
        void freeTree(TrieNode* root) {
            for(int i = 0; i < MAX_CHARS; i++) {
                if((*root)[i] != NULL) {
                    freeTree((*root)[i]);
                }
            }
            delete root;
        }
        
        TrieNode* root;
};

class WordDictionary {
    public:
      void addWord(string word) {
          tree.put(word);
      }
      
      bool search(string word) {
          return tree.search(word);
      }
      
    private: 
    TrieTree tree;
};